Create "ceph-nvmeof" CLI command to use nvmeof-cli

VallariAg opened the pull request:
Add "ceph-nvmeof" cli command to package this as python cli tool instead of using via containers.

How to build it:
```
git clone https://github.com/ceph/ceph-nvmeof.git
cd ceph-nvmeof/
dnf install -y python3-pip
python3 -m venv venv
source venv/bin/activate
pip install -U pip setuptools
pip install pdm==2.7.4
pdm "sync" -v --no-isolation --no-self --no-editable
pdm run protoc
pdm build

pip install dist/ceph_nvmeof-1.2.15-py3-none-any.whl
ceph-nvmeof
```

Usage:
```
[root@singleESXiS1in ~]# pip3 install ceph-nvmeof
[root@singleESXiS1in ~]# ceph-nvmeof --server-address $NVMEOF_DEFAULT_GATEWAY_IP_ADDRESS --server-port $NVMEOF_SRPORT --format json subsystem list
{
    "error_message": "Success",
    "status": 0,
    "subsystems": []
}
[root@singleESXiS1in ~]# ceph-nvmeof  --server-address $NVMEOF_DEFAULT_GATEWAY_IP_ADDRESS --server-port $NVMEOF_SRPORT subsystem add --subsystem nqn.2016-06.io.spdk:cnode1
Adding subsystem nqn.2016-06.io.spdk:cnode1: Successful
[root@singleESXiS1in ~]# ceph-nvmeof --server-address $NVMEOF_DEFAULT_GATEWAY_IP_ADDRESS --server-port $NVMEOF_SRPORT --format json subsystem list
{
    "error_message": "Success",
    "subsystems": [
        {
            "nqn": "nqn.2016-06.io.spdk:cnode1",
            "enable_ha": true,
            "serial_number": "Ceph36100720827891",
            "model_number": "Ceph bdev Controller",
            "min_cntlid": 1,
            "max_cntlid": 2040,
            "subtype": "NVMe",
            "max_namespaces": 256,
            "namespace_count": 0
        }
    ],
    "status": 0
}
```
Package: https://pypi.org/project/ceph-nvmeof/

VallariAg committed:
pyproject.toml: add "[project.scripts]" section

Add "ceph-nvmeof" cli command to use it as a CLI
instead of containers.

Signed-off-by: Vallari Agrawal <val.agl002@gmail.com>

baum reviewed (approved):
lgtm 🖖

epuertat commented on .github/workflows/testpypi.yml:
```diff
@@ -0,0 +1,40 @@
+# Publish python package at https://test.pypi.org/p/ceph-nvmeof
+# This is for dev releases, and will be pushed to test pypi on all merges to "devel" branch.
+# Usage: `pip install --index-url https://test.pypi.org/simple/ --extra-index-url https://pypi.org/simple ceph-nvmeof==1.2.15`
+name: Upload release to Test PyPI
+on:
+  push:
+    tags:
+      - '*'
+    branches:
+      - 'devel'
```
Given that the test and prod workflows are 99% same, I'd try to merge them into a single one accepting an input flag (`prod`/`test`), or building `prod` packages for `v*.*.*` tags, and `test` otherwise. From experience, maintaining 2 different versions of the (almost) same workflow is counterproductive and leads to mismatches.

epuertat commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,33 @@
+      - uses: actions/checkout@v3
+      - uses: pdm-project/setup-pdm@v3
+        with:
+          python-version: 3.9
+          version: "2.7.4"
```
I'd move all versions as input params of the workflow (defaulting to these same numbers), but that allows to easily build against newer/older Python/PDM versions.

epuertat reviewed (approved):
Great work, Vallari! Now I see the new CLI usage, it'd probably make sense to enable fetching the server address/port from a global env. var too (that will save a lot of repetitive input).
```bash
CEPH_NVMEOF_SERVER_ADDRESS=192.168.1.20
CEPH_NVMEOF_SERVER_PORT=5500

ceph-nvmeof gateway info
```

VallariAg committed:
Add .github/workflows/release.yml

Official releases would be published to
https://pypi.org/project/ceph-nvmeof/
on every tag releases on github.

Dev releases would be published to
https://test.pypi.org/project/ceph-nvmeof/
on every merge to "devel" branch.

Package will override if the version already exists.
For example, if "1.2.15" already exists on test.pypi
and the new commit also has same version (in pyproject.toml),
then that package would be overwritten (by using a bigger
build number).

Signed-off-by: Vallari Agrawal <val.agl002@gmail.com>

VallariAg committed:
README: add instructions to install ceph-nvmeof cli

Signed-off-by: Vallari Agrawal <val.agl002@gmail.com>

VallariAg committed:
control/cli.py: read server-address/port defaults from env

"--server-address" defaults to env variable
CEPH_NVMEOF_SERVER_ADDRESS (otherwise "localhost").

"--server-port" defaults to env variable
CEPH_NVMEOF_SERVER_PORT (otherwise 5500).

This is to avoid repetitive inputs. If these
environment variables are set, we can just do:
`ceph-nvmeof gw info`.
(instead of current `ceph-nvmeof --server-address /
$NVMEOF_SERVER_ADDRESS --server-port /
$NVMEOF_SERVER_PORT gw info`)

Signed-off-by: Vallari Agrawal <val.agl002@gmail.com>

VallariAg commented:
@epuertat I have merged the workflows into one - it looks much better now! Thank you!

 Let me know if my solution of reading default server-address/port in `control/cli.py` is okay!

epuertat commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,70 @@
+      # - name: [debug] Set dev version number    # For debugging CI failures (to avoid messing dev releases)
+      #   run: |
+      #     NEW_VERSION="$(pdm show --version).dev$NOW"
+      #     sed -i "s/^version = \".*\"/version = \"${NEW_VERSION}\"/" pyproject.toml
```
If you want to keep some extra debug info, you can either reuse the [Github approach for debugging Actions](https://docs.github.com/en/actions/monitoring-and-troubleshooting-workflows/enabling-debug-logging), or add a new `DEBUG_...` variable.

Otherwise, this code will never run, so I'd remove it.

epuertat commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,70 @@
+        type: choice
+        options:
+          - dev 
+          - prod-and-dev
```
No `prod` alone? What's the point of publishing to both? Usually you either push to the staging env or to the production env. If you want to do both, then you just trigger the runs twice with either value.

epuertat commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,70 @@
+        run: pdm build --config-setting="--build-number=$NOW" --no-sdist
+      - name: Publish package distributions to PyPI
+        run: pdm publish --no-build
+        if: startsWith(github.ref, 'refs/tags/') || (inputs.pypi_repo && inputs.pypi_repo == 'prod-and-dev')
```
I'd simplify this. When do you want each build type to happen?

* the prod package publishing should ONLY happen as part of the Github "Create release" process, which internally may involve creating a tag (but not necessarily). Hence, the best option would be to trigger [the job on `release: published` events](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#release).
  * If a manual prod amendment is required, the `workflow_dispatch` would do the trick.
* Test packages may happen (depending on whether PyPI has some storage quota or not):
  * On-demand (`worklfow_dispath`),
  * Whenever a PR is opened (`pull_request_target` against `devel`),
  * Whenever a PR is merged (`push` against `devel`).

VallariAg replied:
@epuertat, thank you so much for the reviews! Sorry for the delay in updating the PR.
I looked into pypi storage limits and it's 10GB/project: https://pypi.org/help/#:~:text=project%20size%20limit%20(10.0%20GiB%20by%20default

So I have added an input in `workflow_dispatch` called "ref" which takes in which branch/commit to checkout before building. This doesn't cover pull requests made from forks (for which I can add an input to accept "repo" to checkout), but I am not sure who would have permissions to run these manual workflow so we don't want someone to accidentally update the prod package from a fork's branch.
From what I could think, we have few options for open pull requests:
1. Push packages for all pull request (with version `dev<pull_request_id>`) and later go and delete some builds from test pypi UI whenever pushes start failing.
2. Add a note in pull request template about how to create wheel packages for testing.

epuertat commented on control/cli.py:
```diff
@@ -108,13 +108,13 @@ def __init__(self):
             "--server-address",
-            default="localhost",
+            default=(os.getenv('CEPH_NVMEOF_SERVER_ADDRESS') or "localhost"),
             type=str,
             help="Server address",
         )
         self.parser.add_argument(
             "--server-port",
-            default=5500,
+            default=(os.getenv('CEPH_NVMEOF_SERVER_PORT') or 5500),
```
We should briefly document this (in the CLI usage message) or it'll be just a hidden feature :P

epuertat replied:
```suggestion
            default=int(os.getenv('CEPH_NVMEOF_SERVER_PORT') or "5500"),
```

epuertat reviewed (approved):
Great work, @VallariAg ! Thank you for addressing my comments.

Just left a few more comments that might improve the PR.

epuertat reviewed (approved):
LGMT! Great work @VallariAg !
